flink-table/flink-table-planner/src/main/java/org/apache/flink/table/operations/AggregateOperationFactory.java

```diff
@@ -0,0 +1,265 @@
+				} else {
+					if (functionDefinition instanceof AggregateFunctionDefinition) {
+						if (requiresOver(functionDefinition)) {
+							throw new ValidationException(String.format(
+								"OVER clause is necessary for window functions: [%s].",
+								call));
+						}
+					}
+
+					call.getChildren().forEach(child -> child.accept(noChainedAggregates));
+				}
+			} else {
+				call.getChildren().forEach(expr -> expr.accept(this));
+			}
+			return null;
+		}
+
+		private boolean requiresOver(FunctionDefinition functionDefinition) {
+			return ((AggregateFunctionDefinition) functionDefinition).getAggregateFunction().requiresOver();
+		}
+
+		@Override
+		public Void visitLocalReference(LocalReferenceExpression localReference) {
+			failExpression(localReference);
+			return null;
+		}
+
+		@Override
+		public Void visitFieldReference(FieldReferenceExpression fieldReference) {
+			failExpression(fieldReference);
+			return null;
+		}
+
+		@Override
+		protected Void defaultMethod(Expression expression) {
+			return null;
+		}
+
+		private void failExpression(Expression expression) {
+			if (!availableGroupings.contains(expression)) {
+				throw new ValidationException(String.format("expression '%s' is invalid because it is neither" +
+						" present in group by nor an aggregate function", expression));
+			}
+		}
+	}
+
+	private class ValidateDistinct extends ApiExpressionDefaultVisitor<Void> {
+
+		@Override
+		public Void visitCall(CallExpression call) {
+			if (call.getFunctionDefinition() == BuiltInFunctionDefinitions.DISTINCT) {
+				throw new ValidationException("It's not allowed to use an aggregate function as " +
+					"input of another aggregate function");
+			} else if (call.getFunctionDefinition().getType() != AGGREGATE_FUNCTION) {
+				throw new ValidationException("Distinct operator can only be applied to aggregation expressions!");
+			} else {
+				call.getChildren().forEach(child -> child.accept(noChainedAggregates));
+			}
+			return null;
+		}
+
+		@Override
+		protected Void defaultMethod(Expression expression) {
+			return null;
+		}
+	}
+
+	private class NoChainedAggregates extends ApiExpressionDefaultVisitor<Void> {
+
+		@Override
+		public Void visitCall(CallExpression call) {
+			if (call.getFunctionDefinition().getType() == AGGREGATE_FUNCTION) {
+				throw new ValidationException("It's not allowed to use an aggregate function as " +
+					"input of another aggregate function");
+			}
+			return null;
+		}
+
+		@Override
+		protected Void defaultMethod(Expression expression) {
+			return null;
+		}
+	}
+
+	private class GroupingExpressionValidator extends ApiExpressionDefaultVisitor<Void> {
+		@Override
+		protected Void defaultMethod(Expression expression) {
+			TypeInformation<?> groupingType = expressionBridge.bridge(expression).resultType();
+
+			if (!groupingType.isKeyType()) {
+				throw new ValidationException(String.format("expression %s cannot be used as a grouping expression " +
```

Comment:
nit: I know we should not modify too much of the existing logic but I would fix the exception messages in this step. Like starting with upper case.